File: download_cleaner/go-version/download_cleaner.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

func main() {
	// If there is a directory passed as argument, use it first
	downloadsPath := ""
	if len(os.Args) > 1 {
		downloadsPath = os.Args[1]
		fmt.Printf("Using directory %s\n", downloadsPath)
	} else {
		// 1. Get the user's home directory
		homeDir, err := os.UserHomeDir()
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error al obtener el directorio de inicio: %v\n", err)
			return
		}
		// 2. Build the path to the "Downloads" directory
		downloadsPath = filepath.Join(homeDir, "Downloads")
	}

	// Read the files in the Downloads directory
	files, err := os.ReadDir(downloadsPath)
	if err != nil {
		fmt.Println(err)
		return
	}

	// Create a dictionary to store files by extension
	fileExtensionDict := make(map[string][]string)

	for _, file := range files {

		// Skip directories
		if file.IsDir() {
			continue
		}
		// Get the file extension
		key := filepath.Ext(file.Name())
		keyClean, err := strings.CutPrefix(key, ".")
		if err != true {
			fmt.Printf("%s does not have an extension\n", file.Name())
			continue
		}

		// Exclude .DS_Store filename
		if keyClean == "DS_Store" {
			fmt.Println("Excluding .DS_Store filename")
			continue
		}
		ext := strings.ToUpper(keyClean)

		// Add the file to the dictionary
		if _, ok := fileExtensionDict[ext]; !ok {
			fileExtensionDict[ext] = []string{}
		}
		fileExtensionDict[ext] = append(fileExtensionDict[ext], file.Name())
	}

	// Create a directory for each extension in the dictionary
	for ext, files := range fileExtensionDict {
		extPath := filepath.Join(downloadsPath, ext)
		if err := os.Mkdir(extPath, 0777); err != nil {
			fmt.Println(err)
		}

		// Move all files with the current extension and date and permissions to the new directory
		for _, fileName := range files {
			srcPath := filepath.Join(downloadsPath, fileName)
			dstPath := filepath.Join(extPath, fileName)
			if err := os.Rename(srcPath, dstPath); err != nil {
				fmt.Println(err)
			}
		}

		fmt.Printf("Moved all %s files to %s\n", ext, extPath)
	}
}

File: download_cleaner/rust-version/download_cleaner.rs
use std::collections::HashMap;
use std::env;
use std::fs;
use std::path::{Path, PathBuf};

fn main() {
    let mut downloads_path: PathBuf = PathBuf::new();
    // If the user pass a directory as argument, use it first
    if env::args().len() > 1 {
        downloads_path = Path::new(&env::args().nth(1).unwrap()).to_path_buf();
        println!("Using directory {}", downloads_path.display());
    } else {
        // 1. Get the user's home directory
        let home_dir = match env::var("HOME") {
            Ok(dir) => dir,
            Err(_) => {
                eprintln!("Error al obtener el directorio de inicio");
                return;
            }
        };
        // 2. Build the path to the "Downloads" directory
        downloads_path = Path::new(&home_dir).join("Downloads");
    }

    // Read the files in the Downloads directory
    let files = match fs::read_dir(&downloads_path) {
        Ok(entries) => entries,
        Err(err) => {
            println!("{}", err);
            return;
        }
    };

    // Create a dictionary to store files by extension
    let mut file_extension_dict: HashMap<String, Vec<String>> = HashMap::new();

    for file in files {
        let file = match file {
            Ok(entry) => entry,
            Err(_) => continue,
        };

        // Skip directories
        if file.file_type().unwrap().is_dir() {
            continue;
        }

        let file_name = file.file_name().to_string_lossy().to_string();

        // Get the file extension
        let extension = match Path::new(&file_name).extension() {
            Some(ext) => ext.to_string_lossy().to_string(),
            None => {
                println!("{} does not have an extension", file_name);
                continue;
            }
        };

        // Exclude .DS_Store filename
        if extension == "DS_Store" {
            println!("Excluding .DS_Store filename");
            continue;
        }

        let ext = extension.to_uppercase();

        // Add the file to the dictionary
        file_extension_dict
            .entry(ext)
            .or_insert_with(Vec::new)
            .push(file_name);
    }

    // Create a directory for each extension in the dictionary
    for (ext, files) in file_extension_dict {
        let ext_path = downloads_path.join(&ext);

        if let Err(err) = fs::create_dir(&ext_path) {
            println!("{}", err);
        }

        // Move all files with the current extension to the new directory
        for file_name in files {
            let src_path = downloads_path.join(&file_name);
            let dst_path = ext_path.join(&file_name);

            if let Err(err) = fs::rename(&src_path, &dst_path) {
                println!("{}", err);
            }
        }

        println!("Moved all {} files to {}", ext, ext_path.display());
    }
}
